modules/Blockchain/Ethereum/contracts/Marketplace.sol

```diff
@@ -0,0 +1,260 @@
+pragma solidity ^0.4.0;
+
+import './Hub.sol';
+
+import {ERC725} from './ERC725.sol';
+import {HoldingStorage} from './HoldingStorage.sol';
+import {MarketplaceStorage} from './MarketplaceStorage.sol';
+import {SafeMath} from './SafeMath.sol';
+import {ProfileStorage} from "./ProfileStorage.sol";
+import {Profile} from "./Profile.sol";
+
+contract Marketplace {
+    using SafeMath for uint256;
+
+    Hub public hub;
+
+    modifier onlyContracts(){
+        require(hub.isContract(msg.sender),
+            "Function can only be called by contracts!");
+        _;
+    }
+
+    uint256 public paymentStageInterval = 5 minutes;
+
+    constructor(address hubAddress) public {
+        require(hubAddress != address(0));
+        hub = Hub(hubAddress);
+    }
+
+    function setHubAddress(address newHubAddress)
+    public onlyContracts {
+        require(newHubAddress != address(0));
+        hub = Hub(newHubAddress);
+    }
+
+    modifier onlyHolding(){
+        require(msg.sender == hub.getContractAddress("Holding"),
+            "Function can only be called by Holding contract!");
+        _;
+    }
+
+    event PurchaseInitiated(bytes32 purchaseId, address sellerIdentity, address buyerIdentity, uint256 price, bytes32 encodedDataRootHash);
```

Comment:
Should we also return originalDataRootHash?

Reply:
Sure, makes sense